File: 940-fruit-into-baskets/fruit-into-baskets.java
class Solution {
    public int totalFruit(int[] fruits) {
        HashMap<Integer,Integer> mp = new HashMap<>();

        int start=0;
        int end=0;
        int len=0;

        while(end<fruits.length){
            if(!mp.containsKey(fruits[end])){
                mp.put(fruits[end],1);
            }
            else{
                mp.put(fruits[end],mp.get(fruits[end])+1);
            }
            if(mp.size()>2){
                mp.put(fruits[start],mp.get(fruits[start])-1);
                if(mp.get(fruits[start])==0){
                    mp.remove(fruits[start]);
                }
                start++;
            }
            else if(mp.size()<=2){
                len=Math.max(len,end-start+1);
            }
            end++;
        }
        return len;
    }
}

File: 940-fruit-into-baskets/fruit-into-baskets.py
class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        dic = {}

        start=0
        end=0
        length=0

        while end<len(fruits):
            if fruits[end] in dic:
                dic[fruits[end]]+=1
            else:
                dic[fruits[end]]=1

            if len(dic)>2:
                dic[fruits[start]]-=1
                if dic[fruits[start]]==0:
                    del dic[fruits[start]]
                start+=1

            length=max(length,end-start+1) 
            end+=1

        return length